File: Solutions/Python/binary_search_tree_iterator.py
"""
    [MEDIUM]
    173. Binary Search Tree Iterator

    Concepts:
    - binary search tree
    - depth-first search

    Stats:
        Runtime | 5 ms      [Beats 72.25%]
        Memory  | 25.01 MB  [Beats 26.69%]
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.curr = -1
        self.iot = []

        def dfs(curr):
            if not curr:
                return

            dfs(curr.left)
            self.iot.append(curr.val)
            dfs(curr.right)

        dfs(root)
        self.n = len(self.iot)
        

    def next(self) -> int:
        self.curr += 1
        return self.iot[self.curr]

    def hasNext(self) -> bool:
        if self.curr == -1:
            return self.iot[0] is not None

        return self.curr + 1 < self.n

        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: Solutions/Cpp/binary_search_tree_iterator.cpp
/*
    [MEDIUM]
    173. Binary Search Tree Iterator

    Concepts:
    - binary search tree
    - depth-first search

    Stats:
        Runtime | 4 ms      [Beats 54.84%]
        Memory  | 32.15 MB  [Beats 23.51%]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
public:
    int n, idx;
    std::vector<int> iot;

    BSTIterator(TreeNode* root) {
        idx = -1;
        iot = {};
        dfs(root);
        n = iot.size();
    }

    void dfs(TreeNode* curr) {
        if (curr == nullptr)
            return;

        dfs(curr->left);
        iot.push_back(curr->val);
        dfs(curr->right);
    }
    
    int next() {
        return iot.at(++idx);
    }
    
    bool hasNext() {
        if (idx == -1)
            return &iot.front() != nullptr;
        return idx + 1 < n;
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */